Documentation: add a doc of v3 auth

mitake opened the pull request:
It is almost same to Documentation/v2/authentication.md because a
major part of its user interface is shared with the v2 auth. The newly
added doc includes some refinements for the v3 auth.

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+### User `root`
+
+User `root` must be created before auth can be activated. It must have the `root` role and is allowed to change anything inside etcd. The idea behind the `root` user is for administrative purposes: managing other ordinal users and roles.  In addition, the root role is required for other management functionalities including adding/removing/updating a node in a cluster, defragmentation, taking a sn ...
```
The `root` user, which has full access to etcd, must be created before activating authentication. The idea behind the `root` user is for administrative purposes: managing roles and ordinary users.  The `root` user must have the `root` role and is allowed to change anything inside etcd.
?

Is `ordinal` meant to be `ordinary`? I've never seen `ordinal` used this way (aside from a few comments in the etcd code).

mitake replied:
The usage of `ordinal` must be my misunderstanding of English... thanks for pointing! I'll fix other `ordinal`s in other PR.

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+### Role `root`
+
+Role `root` can be granted to any user other than root. Having access via the root role not only allows global read-write access (as was the case before 2.1) but allows modification of the authentication policy and all administrative things, like modifying the cluster membership.
```
The role `root` may be granted to any user, in addition to the root user. A user with the root role has both global read-write access and permission to update the cluster's authentication configuration. Furthermore, the root role grants privileges for general cluster maintenance, including modifying cluster membership, defragmenting the store, and taking snapshots.
?

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+The `user` subcommand for `etcdctl` handles all things having to do with user accounts.
+
+A listing of users can be found with
```
s/with/with:/

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl user list
+```
+
+Creating a user is as easy as
````
Create a new user with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl user add myusername
+```
+
+And there will be prompt for a new password. The password can be supplied from standard input when an option `--interactive=false` is given.
````
s/and there will be/Creating a new user will/

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+And there will be prompt for a new password. The password can be supplied from standard input when an option `--interactive=false` is given.
+
+Roles can be granted and revoked for a user with
```
s/with/with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl user revoke-role myusername bar
+```
+
+We can look at this user with
````
The user's settings can be inspected with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl user get myusername
+```
+
+And the password for a user can be changed with
````
The user's password can be changed with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl user passwd myusername
+```
+
+Which will prompt again for a new password. The password can be supplied from standard input when an option `--interactive=false` is given.
````
s/Which/Changing the password/

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+Which will prompt again for a new password. The password can be supplied from standard input when an option `--interactive=false` is given.
+
+To delete an account, there's always
```
Delete an account with:

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+The `role` subcommand for `etcdctl` handles all things having to do with access controls for particular roles, as were granted to individual users.
+
+A listing of roles can be found with
```
List roles with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl role list
+```
+
+A new role can be created with
````
Create a new role with:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl role add myrolename
+```
+
+A role has no password; we are merely defining a new set of access rights.
````
s/we are merely defining/it merely defines

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+Roles are granted access to a single key or a range of keys.
+
+The range can be specified in a form of [start-key, end-key) where start-key should be smaller than end-key in an alphabetical manner.
```
s/in a form of/as an interval
s/should be smaller than end-key in an alphabetical manner/should be lexically less than end-key, ?

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl role grant-permission myrolename --prefix=true readwrite /pub/
+```
+
+To see what's granted, we can look at the role at any time:
````
s/we can look at the role at any time/look at the role:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl role revoke-permission myrolename /foo/bar
+```
+
+As is removing a role entirely
````
s/entirely/entirely:

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+Password of root:
+```
+
+And enable authentication
````
`Enable authentication:`?

heyitsanthony commented on Documentation/authentication.md:
````diff
@@ -0,0 +1,164 @@
+$ etcdctl --user user:password get foo
+```
+
+The password can also be to be prompted:
````
s/be to be prompted:/be taken from a prompt:/

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+Otherwise, all `etcdctl` commands remain the same. Users and roles can still be created and modified, but require authentication by a user with the root role.
+
+## Using Common Name of TLS cert file
```
`Using TLS Common Name`?

heyitsanthony commented on Documentation/authentication.md:
```diff
@@ -0,0 +1,164 @@
+
+## Using Common Name of TLS cert file
+
+If an etcd server is launched with an option `--client-cert-auth=true`, the field of Common Name (CN) in the client's TLS cert will be used as an etcd user. In this case, the client will not be required the password based authentication.
```
s/an option/the option
s/the client will not be required the password based authentication./the common name authenticates the user and the client does not need a password.

xiang90 commented:
@mitake 

move this doc to doc/op-guide and add a link in section https://github.com/coreos/etcd/tree/master/Documentation#operating-etcd-clusters?

mitake commented:
@heyitsanthony thanks a lot for correcting! The latest version reflects your comments.
@xiang90 I also moved the doc and added a link to `Documentation/docs.md`, could you take a look?

heyitsanthony commented on Documentation/docs.md:
```diff
@@ -37,6 +37,7 @@ Administrators who need to create reliable and scalable key-value stores for the
  - [Performance][performance]
  - [Versioning][versioning]
  - [Supported platform][supported_platform]
+ - [Authentication][authentication]
```
put between security and monitoring? seems like it belongs closer to the setup stuff

mitake replied:
OK, I'll move it.

heyitsanthony commented:
lgtm, maybe move the docs.md authentication link up the list if it makes sense. Thanks!

mitake commented:
@heyitsanthony reordered the list, could you take a look?

heyitsanthony commented:
lgtm

xiang90 commented:
@mitake Thank you!
